### ISSUE TITLE
Passing values as parameters to a useStore call

### ISSUE THREAD

**Issue by cefn** [posva/pinia#1061]
### What problem is this solving

Related to https://github.com/vuejs/pinia/discussions/826 there was an open question what problem would be solved by allowing values to be passed to `useStore`. For my case I need different useStore calls to get different configs to facilitate testing. It was proposed to use an `init()` action, but this introduces unnecessary complexity as described under `Workarounds` below.

No arguments are passed from `useStore` to my `defineStore` setup function. There is no access to the `Pinia` instance from the setup function. There is no `this` definition for the setup function. If any of these constraints were relaxed, there would be a way for me to pass information to the setup function and fulfil distinct testing scenarios without making production code unnecessarily complex.

### Motivating example

In my production pinia Store I have a globally-defined HTTP endpoint URL (set by an env variable) that pinia business logic uses to populate local data within Store state. Asynchronous round-trips to fetch data are triggered by watches of UI state that are defined in the `defineStore` 'setup' function.

In tests, this endpoint is fulfilled dynamically by a dedicated backend server composed specifically for the conditions of each test. This endpoint will come up on a dynamically assigned available port, and multiple tests are expected to execute concurrently.

Setting a global endpoint config variable (e.g. setting `process.env` or mocking a config module) on a per-test basis makes running parallel tests flaky without close coupling to the moment the config variable is interrogated or without [some kind of lock](https://cefn.com/lauf/api/modules/_lauf_lock.html) in place. Race conditions could mean a test sometimes launches with the wrong endpoint. Wrapping tests in a lock would end up causing all tests to run serially. This seems far from elegant.

Better would be some kind of config 'scoping' with the obvious example being options passed to the routine that initialises the business logic. However, passing a zero-arg setup function `() => SS` to `defineStore` (which defines the business logic through watches) then becomes a problem, since there is no way to initialise each `useStore` differently.

### Proposed solution

I am not wedded to any particular solution, but proposing one might be a good way to explain the problem too. 

In one approach the `() => SS` in `defineStore` could have the `Pinia` as a single argument ( `(pinia:Pinia) => SS` ). We could type pinia.env as a dedicated `Readonly<Record<string,any>>` populated via `createPinia(env:Readonly<Record<string,unknown>>)` ?

This leaves people the option to ignore the argument (many stores won't need it) but where initialisation is needed, you can declare the pinia argument and interrogate pinia.env in a typesafe manner. Then anything which needs to be globally available within a single pinia context can be provided and consumed on store initialisation.

#### `init()` action

In https://github.com/vuejs/pinia/discussions/826#discussioncomment-1690020 it was proposed to workaround this problem by using an init() action.

Adopting the convention of an `init()` action has negative consequences. It means every business logic path has to handle the case that the store is not yet initialised, and that stores don't get a chance to 'hard fail on launch if they don't have everything they need. 

Having a possibly-uninitialised-state is a completely artificial problem, since the endpoint URL is always globally defined in every production and test environment before any store code runs. The only reason for there being an 'uninitialised' state is being unable to pass information in the context of a `useStore` call.

#### `launch(store, options)` function

A workaround could be to define a function like `launch(store, options)` which takes responsibility for configuring watches, and can take extra `options` arguments. 

Unfortunately, this approach means that also operations which should ideally be store actions can't be defined that way. For example a `refreshCache(id)` function would also need access to the endpoint config. In the end, we end up with all such operations defined like... refreshCache(store, options, id) and we have broken the model of a self-contained store with well-defined action methods like `store.refreshCache(id)`.

#### Dedicated useEnvStore

This is probably the best workaround and a reason to close the ticket (rubber ducking the issue has helped!). I will prototype the approach and prove it worked.

If env variables are needed, why not create a `useEnvStore` which (implicitly) shares the same pinia when invoked from within the `defineStore` setup function. This could be populated in advance for a given test scenario or be mocked by [createTestingPinia()](https://pinia.vuejs.org/api/modules/pinia_testing.html) . The useEnvStore() for each `test` pinia will have its state populated differently before the first `useStore` runs its setup function. The store logic can be coded to 'hard fail' if it doesn't have the endpoint config it needs.

**Comment by cefn** [comment]
# Dedicated useEnvStore

The approach of using a dedicated 'env variable store' (which exploits the fact any useStore within a store setup function uses that store's own Pinia instance) is proven in this standalone test...

```typescript
import { reactive, watch } from "vue";
import { createPinia, defineStore } from "pinia";
import fetch from "cross-fetch";

interface PrimaryStore {
  env: {
    dogApiUrl?: string;
  };
}

interface SecondaryStore {
  images: {
    dog?: string;
  };
}

const usePrimaryStore = defineStore(
  "primary",
  (): PrimaryStore => ({ env: reactive({}) })
);

const useSecondaryStore = defineStore("secondary", (): SecondaryStore => {
  const {
    env: { dogApiUrl },
  } = usePrimaryStore();

  const images: SecondaryStore["images"] = reactive({});

  if (!dogApiUrl) {
    throw new Error("Secondary store requires dogApiUrl");
  }
  const refreshImageUrl = async () => {
    const response = await fetch(dogApiUrl);
    if (response.status === 200) {
      const data = (await response.json()) as { message: string };
      images.dog = data.message;
    }
  };

  refreshImageUrl();

  return {
    images,
  };
});

describe("Pinia useEnvStore experiment", () => {
  test("useSecondaryStore implictly reads usePrimaryStore from same Pinia", async () => {
    const pinia = createPinia();
    const primaryStore = usePrimaryStore(pinia);
    primaryStore.$state.env.dogApiUrl =
      "https://dog.ceo/api/breeds/image/random";

    const secondaryStore = useSecondaryStore(pinia);

    const dog = await new Promise((resolve) =>
      watch(() => secondaryStore.images.dog, resolve)
    );

    expect(typeof dog).toBe("string");
    expect((dog as string).startsWith("https://images.dog.ceo/"));
  });
});

```

**Comment by kawazoe** [comment]
What are the advantages of this solution when compared to only wrapping the "defineStore" in a function call?

```typescript
const useStore = (dogApiUrl: string) => defineStore("secondary", (): SecondaryStore => {
  const images: SecondaryStore["images"] = reactive({});

  if (!dogApiUrl) {
    throw new Error("Secondary store requires dogApiUrl");
  }
  const refreshImageUrl = async () => {
    const response = await fetch(dogApiUrl);
    if (response.status === 200) {
      const data = (await response.json()) as { message: string };
      images.dog = data.message;
    }
  };

  refreshImageUrl();

  return {
    images,
  };
});

describe("Pinia useEnvStore experiment", () => {
  test("useStore is parametrized", async () => {
    const pinia = createPinia();

    const store = useStore("https://dog.ceo/api/breeds/image/random")(pinia);

    const dog = await new Promise((resolve) =>
      watch(() => store.images.dog, resolve)
    );

    expect(typeof dog).toBe("string");
    expect((dog as string).startsWith("https://images.dog.ceo/"));
  });
});
```

**Comment by cefn** [comment]
I'll focus on answering your question `What are the advantages of this solution` by describing the advantages of the 'env store' solution described above. I am not sure my original solution, (to allow value-passing to useStore), actually resolved the underlying problem for all cases, so I don't want to defend it.

## DEPENDENCY INJECTION

So the underlying problem I think the `env store` approach resolves is dependency injection. The secondary store needs a URL from somewhere but neither the store or users of the store should know how to fulfil that dependency.

Instead, the upstream URL should be an environment variable on a ['twelve factor' pattern](https://12factor.net/backing-services). In the case of a test, the URL may be fully transient as per my original case, in the staging or live case, it should be set to the backing service for that deployment environment. It would be bad design for store code, (or production code that uses the store), to have config that specifies the backing service URL, or to have logic that calculates the URL (and hence embeds deep knowledge of environment variants).

## PROBLEM WITH THE ORIGINAL

My original proposal for `useStore` to pass parameters is not ideal. While an optional parameter could be passed to specialise the store (to allow it to be overridden the for a test suite) this falls back to a pattern in production (where no parameters are passed) where a store has the logic to calculate the URL built-in. The production behaviour violates the separation of concerns of the dependency injection pattern. It also creates weird artefacts, for example, that the second caller of useStore might either have their config parameters ignored, or reset some store values by mistake.

## ENV STORE

When I considered making a feature request that defineStore gets passed a reference to its `pinia`, (so I could pass config as properties on `pinia`).  I realised what I had overlooked. It already implicitly has access to a `pinia` through other `useXXXStore` calls, since they will share a `pinia`. So it is cleaner to create typed config data as a store itself.

Using a `pinia` instance and an 'env store' fulfils the criterion of clean dependency injection to my view. The pinia instance represents the store instance's 'environment' and multiple stores can be brought into existence with a different environment, even in the same javascript interpreter, and without ever changing store code, or changing the code of those using the store. Mechanisms already exist for components within a single Vue app to share a pinia instance as their default 'environment'.

## SUGGESTED APPROACH from @kawazoe

Given my intent, the pattern you shared is worse for separation of concerns, every user of the store has to know the URL in order to use the store, and not just test suite authors who want to override it.

## TLDR;

My aim is that a Vue component can treat the store (and its backing services) as a black box, and passing configuration values is therefore a poor pattern. Sharing values via another store in the same `pinia` aligns well with the assumptions of the framework and allows dependency injection through decoupled, environment-specific code that sets those values potentially in multiple different ways for 'secondary' stores to retrieve.

**Comment by kawazoe** [comment]
> Given my intent, the pattern you shared is worse for separation of concerns, every user of the store has to know the URL in order to use the store, and not just test suite authors who want to override it.

Not really. You can still create a global `useStore` function like you would with `defineStore` normally. I don't see why there would be inherent coupling issues when it comes to using higher-order functions. My example focused on the test side of things, but it is trivial to expose a pre-configured `useStore` function with production parameters:

```typescript
const defineApiStore = (endpoint) =>  defineStore(...);

// For the app
export const useStore = defineApiStore('a.prod/endpoint');

// In tests
const useTestStore = defineApiStore('a.test/endpoint');
```

The only downside is that you cannot change the value after the app is created. In that case then, yes, having a store for your app configuration is definitely the way to go.

**Comment by cefn** [comment]
My intended meaning of [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection#:~:text=The%20code%20that%20passes%20the%20service%20to%20the%20client%20is%20called%20the%20injector.%20Instead%20of%20the%20client%20specifying%20which%20service%20it%20will%20use,%20the%20injector%20tells%20the%20client%20what%20service%20to%20use.) is well summarised in wikipedia at the time of writing like...

>  The code that passes the service to the client is called the injector. Instead of the client specifying which service it will use, the injector tells the client what service to use.

To fulfil this pattern, the signature for composing a store in a component in production would be...

```typescript
const store = useStore()
```

And the signature under testing would be...

```typescript
const store = useStore()
```

...because it would be the same code. 

The code you describe as `trivial to expose a pre-configured useStore`...

```typescript
// For the app
export const useStore = defineApiStore('a.prod/endpoint');

// In tests
const useTestStore = defineApiStore('a.test/endpoint');
```

...is the opposite of the dependency injection pattern, in which you need to compose the dependencies explicitly when using it, rather than just using some reference and trusting to an 'injector' to have decided how the reference was composed, hence allowing separation of concerns.

On the [conventional import pattern](https://pinia.vuejs.org/introduction.html#basic-example) of `useStore`...
```typescript
import { useCounterStore } from '@/stores/counter'
```
...you can't have the case that `useStore` needs to be differently launched in the test case, else the test configuration bleeds into the production components. If I have `useStore` as an import from a production package, the cat is already out of the bag and my component already has a direct reference to the production useStore. Changing store contents via mocking a pinia `useStore` during tests instead of just populating the store differently seems mad.

There is potential to use prop drilling or provide/inject to pass a different store to test components, but that is not the conventional `pinia` pattern and is an unnecessary pain I want to avoid. The feature pinia **_already provides_**  is a manageable and isolated instance of typed state for a component tree that is available to any component by importing `useStore` and calling it in `setup()`. So if the state has to be different in a test case, I want that control to be through pinia.

I don't see how your proposal fulfils the requirements of dependency injection. I still have to invent a mechanism for the `useStore` reference to be aliased for `useTestStore` during a test. My proposal explicitly solves this, by relying on the `pinia` context to define distinctive values for some store invocations, hence effectively using the `pinia` instance as an injector itself.
